File: jbake-core/src/main/java/org/jbake/parser/YamlEngine.java
package org.jbake.parser;

import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.model.DocumentModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Map;

public class YamlEngine extends MarkupEngine {

    private static final Logger LOGGER = LoggerFactory.getLogger(YamlEngine.class);

    public static final String JBAKE_PREFIX = "jbake-";

    /**
     * Parses the YAML file and ensures the output is always a Map.
     *
     * @param file
     * @return
     */
    @SuppressWarnings("unchecked")
    private DocumentModel parseFile(File file) {
        DocumentModel model = new DocumentModel();
        Yaml yaml = new Yaml();
        try (InputStream is = new FileInputStream(file)) {
            Object result = yaml.load(is);
            if (result instanceof List) {
                model.put("data", result);
            } else if (result instanceof Map) {
                model.putAll((Map<? extends String, ?>) result);
            } else {
                LOGGER.warn("Unexpected result [{}] while parsing YAML file {}", result.getClass(), file);
            }
        } catch (IOException e) {
            LOGGER.error("Error while parsing YAML file {}", file, e);
        }
        return model;
    }

    @Override
    public DocumentModel parse(JBakeConfiguration config, File file) {
        return parseFile(file);
    }

    /**
     * This method implements the contract allowing use of Yaml files as content files
     *
     * @param context the parser context
     */
    @Override
    public void processHeader(final ParserContext context) {
        DocumentModel fileContents = parseFile(context.getFile());
        DocumentModel documentModel = context.getDocumentModel();

        fileContents.keySet().forEach(key -> {
            if (hasJBakePrefix(key)) {
                String pKey = key.substring(6);
                documentModel.put(pKey, fileContents.get(key));
            } else {
                documentModel.put(key, fileContents.get(key));
            }
        });
    }

    /**
     * This method implements the contract allowing use of Yaml files as content files. As such there is
     * no body for Yaml files so this method just sets an empty String as the body.
     *
     * @param context the parser context
     */
    @Override
    public void processBody(ParserContext context) {
        context.setBody("");
    }

    private boolean hasJBakePrefix(String key) {
        return key.startsWith(JBAKE_PREFIX);
    }
}

File: jbake-core/src/main/java/org/jbake/parser/YamlEngine.kt
package org.jbake.parser

import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import org.jbake.util.Logging.logger
import org.jbake.util.warn
import org.slf4j.Logger
import org.yaml.snakeyaml.Yaml
import java.io.File
import java.io.FileInputStream
import java.io.IOException

class YamlEngine : MarkupEngine() {

    /**
     * Parses the YAML file and ensures the output is always a Map.
     */
    private fun parseFile(file: File): DocumentModel {
        val model = DocumentModel()
        try {
            FileInputStream(file).use { inputStream ->
                when (val result = Yaml().load<Any>(inputStream)) {
                    is MutableList<*> -> model.put("data", result)
                    is MutableMap<*, *> -> {
                        @Suppress("UNCHECKED_CAST")
                        val map = result as? Map<String, Any> ?: emptyMap()
                        model.putAll(map)
                    }
                    else -> log.warn { "Unexpected result [${result.javaClass}] while parsing YAML file $file" }
                }
            }
        }
        catch (e: IOException) { log.error("Error while parsing YAML file $file", e) }
        return model
    }

    override fun parse(config: JBakeConfiguration, file: File): DocumentModel = parseFile(file)

    /**
     * Implements the contract allowing use of Yaml files as content files
     */
    override fun processHeader(context: ParserContext) {
        val fileContents = parseFile(context.file)

        for (key in fileContents.keys) {
            val key_ = if (hasJBakePrefix(key)) key.substring(6) else key
            context.documentModel[key_] = fileContents.get(key) ?: continue
        }
    }

    /**
     * Implements the contract allowing use of Yaml files as content files.
     * As such there is no body for Yaml files, it just sets an empty String as the body.
     */
    override fun processBody(parserContext: ParserContext) {
        parserContext.body = ""
    }

    private fun hasJBakePrefix(key: String): Boolean = key.startsWith(JBAKE_PREFIX)

    private val JBAKE_PREFIX: String = "jbake-"
    private val log: Logger by logger()
}
